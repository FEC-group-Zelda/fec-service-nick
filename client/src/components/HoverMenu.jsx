import React from "react";
import HoverDiv from "../styled_components/hover_menu_style.js";
import Button_Clear from "../styled_components/clear_button.js";

const HoverMenu = props => (
  <HoverDiv>
    <div id="next-bar">
      Next Up <Button_Clear>Clear</Button_Clear>
    </div>

    {props.playList.map((song, id) => {
      return (
        <div key={id}>
          <h4>Track Number</h4>
          <p>{song.id}</p>
          <h4>Title</h4>
          <p>{song.music_title}</p>
          <h4>Artist</h4>
          <p>{song.artist_name}</p>
          <h4>Album Cover</h4>
          <img src={song.album_cover} width="27" height="27" />
        </div>
      );
    })}
  </HoverDiv>
);

export default HoverMenu;
