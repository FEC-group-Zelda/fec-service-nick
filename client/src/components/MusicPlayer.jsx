import React from "react";
import ReactDOM from "react-dom";
import $ from "jquery";
import HoverMenu from "./HoverMenu.jsx";

class MusicPlayer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: "", // currently playing
      playable: false, // set true/false based on loading of src
      playList: [],
      paused: true, // change to false if i want to autoplay on load
      trackNumber: 0,
      displayList: false,
      displayVolume: false,
      seekValue: 0,
      muted: true,
      volume: 0.5
    };
    this.audioRef = React.createRef();
    this.progressRef = React.createRef();
    this.volumeRef = React.createRef();
    this.volumeControlRef = React.createRef();
  }
  // create refs for audio/progress/start-time/end-times/vol-control
  componentDidMount() {
    $.ajax({
      type: "GET",
      url: "http://localhost:4000/getList",
      success: data => {
        console.log(" this is data from client ", data);
        this.setState({
          playList: data,
          selected: data[0].music_url
        });
      }
    });
  }
  //===============PREVIOUS BUTTON
  previous() {
    console.log("clicked previous button");
    var trackNum;
    if (this.state.trackNumber === 0) {
      trackNum = this.state.playList.length - 1;
    } else {
      trackNum = this.state.trackNumber - 1;
    }
    this.changeAudioSource(this.state.playList[trackNum].music_url, trackNum);
    return this.audioRef.current.paused ? this.pauseAudio() : this.playAudio();
  }

  //=================HELPER FUNCTIONS

  playAudio() {
    this.audioRef.current.volume = this.state.volume;
    console.log(this.audioRef.current.volume);
    if (this.state.playable) {
      return this.audioRef.current.play();
    } else {
      this.playAudio();
    }
  }
  pauseAudio() {
    this.setState({ paused: true });
    console.log("audio is paused");
    return this.audioRef.current.pause();
  }

  changeAudioSource(src, trackNum) {
    if (trackNum === undefined) {
      trackNum = this.state.trackNumber;
    }
    this.setState({
      selected: src,
      playable: false,
      trackNumber: trackNum
    });
    console.log("changing audio source");
  }
  //=================EventHandler
  canPlay() {
    this.setState({ playable: true });
    if (!this.state.paused) {
      this.audioRef.current.play();
    }
    console.log("audio source finished loading");
  }
  //=================PLAY/PAUSE BUTTON

  play_pause() {
    // var length = this.audioRef.current.duration;
    // var currentTime = this.audioRef.current.currentTime;
    // console.log(currentTime, length);

    // re-usability problem .. only one ID with audio so it should be fine.. for now.
    console.log("play/pause ", this.audioRef.current.paused);
    // console.log(myAudio.duration);

    this.setState({ paused: !this.state.paused });
    return this.audioRef.current.paused ? this.playAudio() : this.pauseAudio();
  }

  //=====================NEXT BUTTON
  next() {
    var trackNum;
    if (this.state.trackNumber === this.state.playList.length - 1) {
      trackNum = 0;
    } else {
      trackNum = this.state.trackNumber + 1;
    }
    console.log("clicked next button");
    console.log(this.state.trackNumber);
    this.changeAudioSource(this.state.playList[trackNum].music_url, trackNum);
    // this.setState({
    //   // selected: this.state.playList[this.state.trackNumber + 1].music_url,
    //   trackNumber: this.state.trackNumber + 1
    // });
    console.log(this.audioRef.current.duration / 60);
    // myAudio.load();
    console.log(this.audioRef.current.paused);
    return this.audioRef.current.paused ? this.pauseAudio() : this.playAudio();
  }
  //========================SHUFFLE BUTTON
  shuffle() {
    console.log("clicked shuffle button");
    var playList = this.state.playList;
    var shuffle = playList.sort(func);
    function func(a, b) {
      return 0.5 - Math.random();
    }
    console.log(shuffle);

    this.changeAudioSource(
      this.state.playList[this.state.trackNumber].music_url
    );
    this.setState({
      playList: shuffle
      // selected: this.state.playList[this.state.trackNumber].music_url
    });
    // console.log(this.state.trackNumber);
    // console.log(this.state.playList);
    // console.log(this.state.selected);
    return this.audioRef.current.paused
      ? this.audioRef.current.play()
      : this.audioRef.current.pause();
  }
  //=========================REPEAT BUTTON
  repeat() {
    console.log("clicked repeat button");
    if (!this.audioRef.current.loop) {
      this.audioRef.current.loop = true;
    } else {
      this.audioRef.current.loop = false;
    }
  }
  // =========================SHOW PLAYLIST
  showList() {
    this.setState({ displayList: !this.state.displayList });
  }
  // //==========================PROGRESS BAR
  timeUpdate() {
    var length = this.audioRef.current.duration;
    var currentTime = this.audioRef.current.currentTime;

    var minutes = Math.floor(length / 60),
      secondsInt = length - minutes * 60,
      secondsStr = secondsInt.toString(),
      seconds = secondsStr.substr(0, 2);

    if (seconds < 10) {
      seconds = seconds * 10;
    } else {
      seconds;
    }
    var endTime = minutes + ":" + seconds; //<==== total time in correct form

    var currentHour = parseInt(currentTime / 3600) % 24,
      currentMinute = parseInt(currentTime / 60) % 60,
      currentSecond = currentTime % 60,
      current_SecondsFixed = currentSecond.toFixed(),
      current_Time =
        (currentMinute < 10 ? "0" + currentMinute : currentMinute) +
        ":" +
        (current_SecondsFixed < 10
          ? "0" + current_SecondsFixed
          : current_SecondsFixed);

    //onTimeUpdate
    this.setState({ seekValue: Math.floor((currentTime * 100) / length) });
    // state?
  }
  progressBarChange() {
    console.log(this.progressRef.current.value);
    var progressbar = this.progressRef;
    console.log(progressbar);
    var currentTime = this.audioRef.current.currentTime;
    progressbar.current.value = currentTime;
    progressbar.addEventListener("click", this.seek);
    progressbar.addEventListener("drag", this.seek);
  }
  seek(e) {
    e.persist();
    var myAudio = this.audioRef;
    var progressbar = this.progressRef;
    var percent = e.nativeEvent.offsetX / e.target.offsetParent.offsetWidth;
    // var percent = e.target.value / e.target.max;
    console.log(progressbar.current.value);
    console.log(`this is percent`, percent);

    myAudio.current.currentTime = percent * myAudio.current.duration;
    progressbar.current.value = percent;
  }

  //======================VOLUME BAR
  volumeBar(e) {
    console.log(this.audioRef.current.volume);
    this.setState({ volume: e.currentTarget.value / 100 });
    this.audioRef.current.volume = e.currentTarget.value / 100;
  }
  //====================SHOW VOLUME BAR ON CLICK
  displayVolume() {
    this.setState({ displayVolume: !this.state.displayVolume });
  }

  muteVolume() {
    this.setState({ muted: !this.state.muted });
    console.log(this.audioRef.current.muted);
    return (this.audioRef.current.muted = !this.audioRef.current.muted);
  }
  render() {
    return (
      <div id="mini-player">
        <audio
          ref={this.audioRef}
          src={this.state.selected}
          volume={this.state.volume}
          preload="auto"
          onEnded={this.next.bind(this)}
          onCanPlay={this.canPlay.bind(this)}
          onTimeUpdate={this.timeUpdate.bind(this)}
        ></audio>
        <div id="previous-div">
          {" "}
          <button
            className="previous-button"
            onClick={this.previous.bind(this)}
          >
            Previous
          </button>
        </div>
        <div id="play-pause-div">
          {" "}
          <button
            className="playPause-button"
            onClick={this.play_pause.bind(this)}
          >
            {this.state.paused ? "Play" : "Pause"}
          </button>
        </div>

        <div id="next-div">
          {" "}
          <button className="next-button" onClick={this.next.bind(this)}>
            Next
          </button>
        </div>

        <div id="shuffle-div">
          {" "}
          <button className="shuffle-button" onClick={this.shuffle.bind(this)}>
            Shuffle
          </button>
        </div>
        <div id="repeat-div">
          <button className="repeat-button" onClick={this.repeat.bind(this)}>
            Repeat
          </button>
        </div>
        <img
          src={
            this.state.playList.length
              ? this.state.playList[this.state.trackNumber].album_cover
              : ""
          }
          width="52"
          height="52"
        />
        {this.state.playList.length ? (
          <p>
            {this.state.playList[this.state.trackNumber].music_title}
            <small> by </small>
            {this.state.playList[this.state.trackNumber].artist_name}
          </p>
        ) : (
          ""
        )}
        <small id="start-time"></small>
        <span id="seek-container">
          <div id="progress-div">
            <progress
              ref={this.progressRef}
              value={this.state.seekValue || 0}
              max="100"
              onClick={this.seek.bind(this)}
            ></progress>
          </div>
        </span>
        <small id="end-time"></small>

        <div ref={this.volumeRef}>
          <button
            onMouseOver={this.displayVolume.bind(this)}
            onClick={this.muteVolume.bind(this)}
            className="volume-button"
          >
            {this.state.muted ? "Volume" : "Muted"}
          </button>
          {this.state.displayVolume && (
            <div>
              {" "}
              <input
                ref={this.volumeControlRef}
                type="range"
                min="0"
                max="100"
                step="1"
                onChange={this.volumeBar.bind(this)}
              ></input>
            </div>
          )}
        </div>
        <div className="playlist-div">
          <button
            onClick={this.showList.bind(this)}
            className="playList-button"
          >
            PlayList
          </button>
          {this.state.displayList && (
            <div>
              {" "}
              <HoverMenu
                playList={this.state.playList}
                selected={this.state.selected}
              />
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default MusicPlayer;
