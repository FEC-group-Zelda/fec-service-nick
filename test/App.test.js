import React from "react";
import { shallow, mount } from "enzyme";
import MusicPlayer from "../client/src/components/MusicPlayer.jsx";
import HoverMenu from "../client/src/components/HoverMenu.jsx";

describe("Examining the syntax of Jest tests", () => {
  it("sums numbers", () => {
    expect(1 + 2).toEqual(3);
    expect(2 + 2).toEqual(4);
  });
});

describe("MusicPlayer", () => {
  test("should render snapshot", () => {
    const wrapper = mount(<MusicPlayer />);
    expect(wrapper).toMatchSnapshot();
  });

  test("song information should render.", () => {
    const songMock = [
      {
        id: "Test #1",
        music_title: "I'm baby",
        artist_name: "Queen",
        album_cover:
          "https://i.kym-cdn.com/photos/images/original/001/627/845/c22.jpg",
        music_url: "some Url#2"
      },
      {
        id: "Test #2",
        music_title: "uh.",
        artist_name: "Spears",
        album_cover:
          "https://i.kym-cdn.com/photos/images/original/001/627/845/c22.jpg",
        music_url: "some URL #2"
      }
    ];

    const example = shallow(<HoverMenu playList={songMock} />);
    expect(example).toMatchSnapshot();
  });

  test("should not render hoverList if displayList is false.", () => {
    const wrapper = shallow(<MusicPlayer />);
    expect(wrapper.find("HoverMenu").length).toEqual(0);
  });

  //on hover should render the menu
});
