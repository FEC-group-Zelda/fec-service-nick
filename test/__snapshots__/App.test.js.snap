// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicPlayer should render snapshot 1`] = `
<MusicPlayer>
  <styled.div>
    <div
      className="sc-fzozJi cwrFor"
    >
      <audio
        onCanPlay={[Function]}
        onEnded={[Function]}
        onTimeUpdate={[Function]}
        preload="auto"
        src=""
        volume={0.5}
      />
      <div
        id="previous-div"
      >
         
        <styled.button
          className="previous-button"
          onClick={[Function]}
        >
          <button
            className="sc-AxhCb WvUlE previous-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/254/254437.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <div
        id="play-pause-div"
      >
         
        <styled.button
          className="playPause-button"
          onClick={[Function]}
        >
          <button
            className="sc-AxiKw bEmyn playPause-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/254/254434.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <div
        id="next-div"
      >
         
        <styled.button
          className="next-button"
          onClick={[Function]}
        >
          <button
            className="sc-AxhUy fpjEso next-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/254/254428.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <div
        id="shuffle-div"
      >
         
        <styled.button
          className="shuffle-button"
          onClick={[Function]}
        >
          <button
            className="sc-AxgMl jxxhRS shuffle-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/2057/2057590.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <div
        id="repeat-div"
      >
        <styled.button
          className="repeat-button"
          onClick={[Function]}
        >
          <button
            className="sc-AxheI BEzGc repeat-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/565/565272.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <img
        height="27"
        id="queen-image"
        src=""
        width="27"
      />
      <small
        id="start-time"
      />
      <span
        id="seek-container"
      >
        <styled.div>
          <div
            className="sc-fzoLsD bRplmR"
          >
            <progress
              max="100"
              onClick={[Function]}
              value={0}
            />
          </div>
        </styled.div>
      </span>
      <small
        id="end-time"
      />
      <div>
        <styled.button
          className="volume-button"
          onClick={[Function]}
          onMouseOver={[Function]}
        >
          <button
            className="sc-AxmLO gJONoK volume-button"
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/727/727269.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
      <div
        className="playlist-div"
      >
        <styled.button
          className="playList-button"
          onClick={[Function]}
        >
          <button
            className="sc-Axmtr jbAStq playList-button"
            onClick={[Function]}
          >
            <img
              height="30%"
              src="https://image.flaticon.com/icons/svg/565/565266.svg"
              width="30px"
            />
          </button>
        </styled.button>
      </div>
    </div>
  </styled.div>
</MusicPlayer>
`;

exports[`MusicPlayer song information should render. 1`] = `
<styled.div>
  <div
    id="next-bar"
  >
    Next Up 
    <styled.button>
      Clear
    </styled.button>
  </div>
  <div
    key="0"
  >
    <h4>
      Track Number
    </h4>
    <p>
      Test #1
    </p>
    <h4>
      Title
    </h4>
    <p>
      I'm baby
    </p>
    <h4>
      Artist
    </h4>
    <p>
      Queen
    </p>
    <h4>
      Album Cover
    </h4>
    <img
      height="27"
      src="https://i.kym-cdn.com/photos/images/original/001/627/845/c22.jpg"
      width="27"
    />
  </div>
  <div
    key="1"
  >
    <h4>
      Track Number
    </h4>
    <p>
      Test #2
    </p>
    <h4>
      Title
    </h4>
    <p>
      uh.
    </p>
    <h4>
      Artist
    </h4>
    <p>
      Spears
    </p>
    <h4>
      Album Cover
    </h4>
    <img
      height="27"
      src="https://i.kym-cdn.com/photos/images/original/001/627/845/c22.jpg"
      width="27"
    />
  </div>
</styled.div>
`;
